### PR TITLE
add Forward boolean to Jet DataFormat

### PR DESCRIPTION
**Pull request by harmonicoscillator**
@mulhearn @jimbrooke @kalanand @apana 
Stage1Layer2 would like to change our jet finder to make sure that central jets have no HF energy. This means that we need to deal with the possibility of two jets at the same eta-phi position, one including HF (a forward jet) and one not including HF (a central jet). 

I think the simplest method for handling this extra required info in our algos is to add a forward member to the l1t::Jet class, which I have done here. This is build and run-tested. Previous uses of the constructor should still be fine.

Is there any reason not to merge this? I had to edit classes_def.xml, wasn't sure what the procedure with that was.

Thanks,
Alex

### CODE REVIEW

**Comment by jimbrooke** [comment]
I am really reluctant to make changes to data formats to solve problems that originate in the algorithms, such as a requirement that central jets have no HF energy.

However, I can perhaps guess why you want to add this bool.  In the legacy system, the central, tau forward jets are transmitted on separate links, so we were able to save some bandwidth by using different ieta encoding schemes for central and forward jets.  If you are using the same ieta encoding for the stage 1 jet output collection, there is a potential ambiguity.  (Central and forward jets can have the same code but correspond to different ieta).

Here is my recommended solution :
- stage 1 emulator use ieta codes between 0 and 21
- ieta codes 0-3 and 18-21 correspond to forward jets (and should be interpreted as such by the upgrade->legacy module)
- if central jets (ieta codes 4-17) should have no HF energy in them, the algorithm ensures this

If you want to change the definition of the central/forward boundary (which is NOT the same as requiring no HF energy in central jets, and would required agreement with the GT to change the ieta scale) then this could be a configurable parameter in both the stage 1 emulator and the upgrade->legacy module.

Cheers,
Jim

**Comment by jimbrooke** [comment]
PS - another way of putting it, I would say this bool is equivalent to :
((0 >= ieta) && (ieta <=3 )) || ((ieta>=18) && (ieta<=21))

**Comment by harmonicoscillator** [comment]
Jim,

Previously we were doing exactly as you suggested. However we now have the
possibility that two jets, one central and one forward, have the same iEta
code (because their centers are at the same location). That is, there could
be two valid jets with the position ieta=4, one of which is made of 12x12
towers and includes HF energy, and another that is made of 8x12 towers and
does not include HF energy. The central jet may still be the highest pT in
the central collection. Therefore just the ieta code is not enough to
distinguish between central and forward jets, so we need extra information.

If it is a pain to add this information to the dataformat I will try to
think of some other way of conveying this information to the GT. Do you
have a suggestion?

Maybe Kalanand can comment further on why this distinction is important.

Alex

**Comment by jimbrooke** [comment]
Hi Alex,

this is not simply an emulator issue, it's a change in the forward jet definition, which would need to be changed in online DB and elsewhere.  Overlapping central/jet boundary was something that was never envisaged in the legacy system, so I expect you will see unexpected consequences in the hardware.  I would not recommend it.  (I take it the legacy GT team are OK with it?  I am surprised).

But if it must be done, for the data formats I would suggest you use the quality flag member, which was intended for general purpose additional flags, rather than add a new member.

Jim

**Comment by kalanand** [comment]
Hi Alex, Jim,

We want to flag these jets with 8x12 towers, so that if needed
we can calibrate them differently. The quality flag would be
sufficient for this purpose, I think. Alex: let's assign quality
==2 for these jets (or, for that matter, any value you like other
than 0 and 1, because 0 is default and we may want to use 1 for
other quality requirements).

Cheers,
Kalanand

**Comment by harmonicoscillator** [comment]
Kalanand,

Okay, I will switch to using the quality flag.

When you say "quality==2", do you mean that I should check the status of
the second bit in the quality word? i.e. (quality & 0x2 != 0) ? Or do you
actually mean "quality == 2"? I don't understand your explanation for not
using 0 and 1, could you elaborate?

Alex

**Comment by jimbrooke** [comment]
Please assign meaning to single quality bits rather than try to use integer codes !

**Comment by kalanand** [comment]
Yes, please ignore my previous suggestion about
integer codes (I got confused with something else).
Any meaningful quality bit should be fine.

Cheers,
Kalanand

**Comment by harmonicoscillator** [comment]
Superseded by https://github.com/cms-l1t-offline/cmssw/pull/12
